storage/innobase/read/read0read.cc

```diff
@@ -262,4 +307,552 @@ void trx_sys_t::clone_oldest_view(ReadViewBase *view) const
+  ulint offset= scn_mgr->scn_offset(index, offsets);
+  trx_id_t id= mach_read_from_6(rec + offset);
+
+  /* If it's scn, direct compare*/
+  if (SCN_Mgr::is_scn(id))
+  {
+    return sees_version(id);
+  }
+
+  /* Trx itself */
+  if (id == creator_trx_id)
+  {
+    return true;
+  }
+
+  if (id < m_up_limit_id)
+  {
+    return true;
+  }
+
+  if (id >= m_low_limit_id)
+  {
+    return false;
+  }
+
+  if (m_committing_ids.find(id) != m_committing_ids.end())
+  {
+    /* Not visible to current view */
+    return false;
+  }
+
+  /* Get SCN from undo log */
+  trx_id_t committing_version= 0;
+  trx_id_t scn=
+      scn_mgr->get_scn(id, index, row_get_rec_roll_ptr(rec, index, offsets),
+                       &committing_version);
+
+  if (committing_version != 0 && committing_version < m_version)
+  {
+    /* Consider such scenario:
+    - active trx: get trx->no = 5
+    - open read view: version = 7
+    - before committing trx completely: not visible
+    - after committing trx: visible because it's deregistered
+      and scn is written to undo (5 < 7)
+
+    Problem: consistent read is broken, so we must
+    record such kind of scn and id */
+    ut_a(scn == TRX_ID_MAX);
+    m_committing_ids.insert(id);
+    m_committing_scns.insert(committing_version);
+
+    ut_a(committing_version >= m_low_limit_no);
+  }
+
+  if (scn == TRX_ID_MAX)
+  {
+    /* Still active */
+    return false;
+  }
+
+  ut_a(scn > 0);
+
+  if (srv_use_scn && !srv_read_only_mode && block != nullptr &&
+      !index->table->is_temporary() && !index->online_log)
+  {
+    /* Attch record to block */
+    scn_mgr->add_lazy_cursor(block, const_cast<rec_t *>(rec), offset, id, scn,
+                             index->table->id);
+  }
+
+  return (sees_version(scn));
+}
+
+Scn_Map::Scn_Map()
+{
+  m_size= SCN_MAP_MAX_SIZE;
+  m_elems= new Elem[m_size];
+  ut_a(m_elems != nullptr);
+}
+
+Scn_Map::~Scn_Map() { delete[] m_elems; }
+
+void run_cleanout_task(void *arg);
+
+SCN_Mgr::CleanoutWorker::CleanoutWorker(uint32_t id)
+{
+  m_id= id;
+  m_pages= new CleanoutArray(CLEANOUT_ARRAY_MAX_SIZE);
+  m_thd= innobase_create_background_thd("SCN cleanout worker");
+  m_task= new tpool::waitable_task(run_cleanout_task, &m_id);
+}
+
+SCN_Mgr::CleanoutWorker::~CleanoutWorker()
+{
+  delete m_pages;
+  destroy_background_thd(m_thd);
+  delete m_task;
+}
+
+void SCN_Mgr::CleanoutWorker::add_page(uint64_t compact_page_id,
+                                       table_id_t table_id)
+{
+  m_pages->add(compact_page_id, table_id);
+}
+
+void SCN_Mgr::CleanoutWorker::take_pages(PageSets &pages)
+{
+  pages.clear();
+
+  uint64_t page_id= 0;
+  table_id_t table_id= 0;
+  while ((m_pages->get(page_id, table_id)))
+  {
+    pages.insert({page_id, table_id});
+  }
+}
+
+SCN_Mgr::SCN_Mgr()
+{
+  m_scn_map= new Scn_Map();
+  m_random_map= new Scn_Map();
+  m_startup_id= 0;
+  m_startup_scn= 0;
+  m_abort= false;
+  m_max_cleanout_threads= srv_cleanout_threads;
```

Comment:
Can we do without `m_max_cleanout_threads`? One variable name `innodb_cleanout_threads` would be more `grep` friendly.

Reply:
Sure.